### PR TITLE
[doc, fpga] Add bitstream_bisect tool to FPGA reference manual

### PR DESCRIPTION
**Pull request by dmcardle**
This PR documents the tool added in #16406. (Some followup tasks remain in #16788.)

### CODE REVIEW

File: doc/rm/ref_manual_fpga.md [hunk 1]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+    --good HEAD~30 \
+    --bad HEAD \
+    --fast-command "./bazelisk.sh test //sw/device/tests:uart_smoketest_fpga_cw310_rom" \
+    --slow-command "./bazelisk.sh test --define bitstream=vivado //sw/device/tests:uart_smoketest_fpga_cw310_rom"
+```
+
+One caveat is that neither `git-bisect` nor `:bitstream_bisect` will help if the FPGA somehow retains state between tests.
+That is, if the test command bricks the FPGA causing future tests to fail, bisection will return entirely bogus results.
+We plan to add a "canary" feature to `:bitstream_bisect` that will abort the bisect when FPGA flakiness is detected (issue [#16788](https://github.com/lowRISC/opentitan/issues/16788)).
+For now, if you suspect this kind of FPGA flakiness, a linear progression from `BAD_COMMIT` to `GOOD_COMMIT` may be the best answer.
```
Comment (marnovandermaas) [inline doc/rm/ref_manual_fpga.md:118]: If bricking is the problem, it's probably better to go the other way around from good to bad right?

File: doc/rm/ref_manual_fpga.md [hunk 2]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+
+Note that the slow command doesn't necessarily have to build a bitstream.
+If you don't have a Vivado license and the test regression is reproducible in Verilator, it could make sense to fall back to the Verilated test.
+Building on the example above, you could replace the slow command with `"./bazelisk.sh test //sw/device/tests:uart_smoketest_sim_verilator"`, and the `:bitstream_bisect` tool would never build any bitstreams.
```
Comment (marnovandermaas) [inline doc/rm/ref_manual_fpga.md:122]:
```suggestion
Building on the example above, you could replace the slow command with `"./bazelisk.sh test //sw/device/tests:uart_smoketest_sim_verilator"` and the `:bitstream_bisect` tool would never build any bitstreams.
```

File: doc/rm/ref_manual_fpga.md [hunk 3]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+
+For more information, consult the `:bitstream_bisect` tool directly!
+
+```sh
+./bazelisk.sh run //util/fpga:bitstream_bisect -- --help
```
Comment (marnovandermaas) [inline doc/rm/ref_manual_fpga.md:127]: Nice, works for me!

**Review by marnovandermaas — APPROVED** [review]
LGTM, just a few nits.

File: doc/rm/ref_manual_fpga.md (hunk 1 above)
Comment (dmcardle) [reply doc/rm/ref_manual_fpga.md]: Good catch! Just wrote it backwards.

File: doc/rm/ref_manual_fpga.md [hunk 4]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+
+To set the stage, let's say you've discovered a test regression.
+The test used to pass on `GOOD_COMMIT`, but now it fails at `BAD_COMMIT`.
+Your goal is to find the *first bad commit*.
+
+In general, a linear search from `GOOD_COMMIT` to `BAD_COMMIT` is one of the slowest ways to find the first bad commit.
+We can save time by testing fewer commits with `git-bisect`, which effectively applies binary search to the range of commits.
+We can save even more time by leveraging the bitstream cache with **[`//util/fpga:bitstream_bisect`](https://github.com/lowRISC/opentitan/tree/master/util/fpga/bitstream_bisect.py)**.
+
+The `:bitstream_bisect` tool is faster than regular `git-bisect` because it restricts itself to cached bitstreams until it can make no more progress.
```
Comment (drewmacrae) [inline doc/rm/ref_manual_fpga.md:99]: nit: `git bisect`

File: doc/rm/ref_manual_fpga.md [hunk 5]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+Building a bitstream is many times slower than running a test (hours compared to minutes), and `git-bisect` has no idea that some commits will be faster to classify than others due to the bitstream cache.
+
+For example, suppose that `//sw/device/tests:uart_smoketest` has regressed sometime in the last 30 commits.
+The following command could easily save hours compared to a naive `git-bisect`:
```
Comment (drewmacrae) [inline doc/rm/ref_manual_fpga.md:103]: nit: `git bisect`

File: doc/rm/ref_manual_fpga.md [hunk 6]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
 
 For more details on the exact operation of the loading flow and how the boot ROM processes incoming data, please refer to the [boot ROM readme](https://github.com/lowRISC/opentitan/tree/master/sw/device/lib/testing/test_rom).
 
-## Bitstreams: Caching and Splicing
+### Accelerating `git-bisect` with the bitstream cache
```
Comment (drewmacrae) [inline doc/rm/ref_manual_fpga.md:89]: nit: `git bisect`

File: doc/rm/ref_manual_fpga.md [hunk 7]
```diff
@@ -86,7 +86,48 @@ I00005 hello_world.c:76] The LEDs show the ASCII code of the last character.
...
+# This will use the fast command to classify commits with cached bitstreams. If
+# the results are ambiguous, it will narrow them down with the slow command.
+./bazelisk.sh run //util/fpga:bitstream_bisect -- \
+    --good HEAD~30 \
+    --bad HEAD \
+    --fast-command "./bazelisk.sh test //sw/device/tests:uart_smoketest_fpga_cw310_rom" \
+    --slow-command "./bazelisk.sh test --define bitstream=vivado //sw/device/tests:uart_smoketest_fpga_cw310_rom"
+```
+
+One caveat is that neither `git-bisect` nor `:bitstream_bisect` will help if the FPGA somehow retains state between tests.
```
[the remark this answers is not in the document]
Comment (dmcardle) [reply doc/rm/ref_manual_fpga.md:115]: Thanks, I'll make that change. The man page is named `git-bisect` so I started using that as a canonical name, but it probably just raises questions for readers.

[the remark this answers is not in the document]
Comment (dmcardle) [reply doc/rm/ref_manual_fpga.md]: Done. Fixed a docstring in bitstream_bisect.py as well.

**Comment by dmcardle** [comment]
I'll merge now, since the one-char docstring change is in the same spirit as your nits, @drewmacrae. Thanks!

**Pull request merged**